esphome/components/remote_base/nexa_protocol.cpp

```diff
@@ -0,0 +1,327 @@
+#include "nexa_protocol.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace remote_base {
+
+static const char *const TAG = "remote.nexa";
+
+static const uint8_t NBITS = 32;
+static const uint32_t HEADER_HIGH_US = 319;
+static const uint32_t HEADER_LOW_US = 2610;
+static const uint32_t BIT_HIGH_US = 319;
+static const uint32_t BIT_ONE_LOW_US = 1000;
+static const uint32_t BIT_ZERO_LOW_US = 140;
+
+static const uint32_t TX_HEADER_HIGH_US = 250;
+static const uint32_t TX_HEADER_LOW_US = TX_HEADER_HIGH_US * 10;
+static const uint32_t TX_BIT_HIGH_US = 250;
+static const uint32_t TX_BIT_ONE_LOW_US = TX_BIT_HIGH_US * 5;
+static const uint32_t TX_BIT_ZERO_LOW_US = TX_BIT_HIGH_US * 1;
+
+void NexaProtocol::one(RemoteTransmitData *dst) const {
+  // '1' => '10'
+  dst->item(TX_BIT_HIGH_US, TX_BIT_ONE_LOW_US);
+  dst->item(TX_BIT_HIGH_US, TX_BIT_ZERO_LOW_US);
+  // ESP_LOGD(TAG, "1 ");
+}
+
+void NexaProtocol::zero(RemoteTransmitData *dst) const {
+  // '0' => '01'
+  dst->item(TX_BIT_HIGH_US, TX_BIT_ZERO_LOW_US);
+  dst->item(TX_BIT_HIGH_US, TX_BIT_ONE_LOW_US);
+  // ESP_LOGD(TAG, "0 ");
+}
+
+void NexaProtocol::sync(RemoteTransmitData *dst) const {
+  dst->item(TX_HEADER_HIGH_US, TX_HEADER_LOW_US);
+  // ESP_LOGD(TAG, "S ");
+}
+
+void NexaProtocol::encode(RemoteTransmitData *dst, const NexaData &data) {
+  dst->set_carrier_frequency(0);
+  ESP_LOGD(TAG, "ENCODE NEXA: device=0x%04X group=%d state=%d channel=%d level=%d", data.device, data.group, data.state,
+           data.channel, data.level);
+
+  // Send SYNC
+  this->sync(dst);
+
+  // ESP_LOGD(TAG, "Device:");
+  // Device (26 bits)
+  for (int16_t i = 26 - 1; i >= 0; i--) {
+    if (data.device & (1 << i))
+      this->one(dst);
+    else
+      this->zero(dst);
+  }
+
+  // ESP_LOGD(TAG, "Group:");
+  // Group (1 bit)
+  if (data.group != 0)
+    this->one(dst);
+  else
+    this->zero(dst);
+
+  // ESP_LOGD(TAG, "State:");
+  // State (1 bit)
+  if (data.state == 2) {
+    // Special case for dimmers...send 00 as state
+    dst->item(TX_BIT_HIGH_US, TX_BIT_ZERO_LOW_US);
+    dst->item(TX_BIT_HIGH_US, TX_BIT_ZERO_LOW_US);
+  } else if (data.state == 1)
+    this->one(dst);
+  else
+    this->zero(dst);
+
+  // ESP_LOGD(TAG, "Channel:");
+  // Channel (4 bits)
+  for (int16_t i = 4 - 1; i >= 0; i--) {
+    if (data.channel & (1 << i))
+      this->one(dst);
+    else
+      this->zero(dst);
+  }
+
+  // ESP_LOGD(TAG, "Level:");
+  // Level (4 bits)
+  if (data.state == 2) {
+    for (int16_t i = 4 - 1; i >= 0; i--) {
+      if (data.level & (1 << i))
+        this->one(dst);
+      else
+        this->zero(dst);
+    }
+  }
+
+  // Send finishing Zero
+  // this->zero(dst);
+  dst->item(TX_BIT_HIGH_US, TX_BIT_ZERO_LOW_US);
+
+  // Send the PAUSE
+  // dst->mark(BIT_HIGH_US);
+  // dst->space(TX_BIT_HIGH_US*50);
+
+  ESP_LOGD(TAG, "ENCODE NEXA: Done");
```

Comment:
This should not really be any higher than VERBOSE
```suggestion
  ESP_LOGV(TAG, "ENCODE NEXA: Done");
```

Reply:
Will do! I used a lot during dev, but it is not needed anymore.